File: frontend/src/lib/api/community.ts
import { createClient } from "@/lib/supabase/client";
import { getApiBaseUrl } from "@/lib/api/client";
import type {
  CreatePostPayload,
  CreateReplyPayload,
  NotificationOut,
  PostDetail,
  PostListResponse,
  PostSummary,
  SortBy,
  UpvoteResponse,
  VoteResponse,
} from "@/types/community";

async function getAccessToken(): Promise<string> {
  const supabase = createClient();
  const {
    data: { session },
  } = await supabase.auth.getSession();
  if (!session?.access_token) {
    throw new Error("Not authenticated");
  }
  return session.access_token;
}

export async function listPosts(opts?: {
  courseCode?: string;
  professorName?: string;
  search?: string;
  sortBy?: SortBy;
  department?: string;
  courseNumber?: string;
  page?: number;
}): Promise<PostListResponse> {
  const token = await getAccessToken();
  const base = getApiBaseUrl();
  const params = new URLSearchParams();
  if (opts?.courseCode) params.set("course_code", opts.courseCode);
  if (opts?.professorName) params.set("professor_name", opts.professorName);
  if (opts?.search) params.set("search", opts.search);
  if (opts?.sortBy) params.set("sort_by", opts.sortBy);
  if (opts?.department) params.set("department", opts.department);
  if (opts?.courseNumber) params.set("course_number", opts.courseNumber);
  if (opts?.page) params.set("page", String(opts.page));

  const res = await fetch(
    `${base}/api/community${params.size ? `?${params}` : ""}`,
    { headers: { Authorization: `Bearer ${token}` } },
  );
  if (!res.ok) throw new Error(`listPosts failed: ${res.status}`);
  return res.json() as Promise<PostListResponse>;
}

export async function getDepartments(): Promise<string[]> {
  const token = await getAccessToken();
  const res = await fetch(`${getApiBaseUrl()}/api/community/departments`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  if (!res.ok) throw new Error(`getDepartments failed: ${res.status}`);
  return res.json() as Promise<string[]>;
}

export async function getPost(postId: string): Promise<PostDetail> {
  const token = await getAccessToken();
  const res = await fetch(`${getApiBaseUrl()}/api/community/${postId}`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  if (!res.ok) throw new Error(`getPost failed: ${res.status}`);
  return res.json() as Promise<PostDetail>;
}

export async function createPost(
  payload: CreatePostPayload,
): Promise<PostSummary> {
  const token = await getAccessToken();
  const res = await fetch(`${getApiBaseUrl()}/api/community`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${token}`,
    },
    body: JSON.stringify(payload),
  });
  if (!res.ok) {
    const detail = await res.json().catch(() => null);
    const msg = detail?.detail ?? `createPost failed: ${res.status}`;
    throw new Error(typeof msg === "string" ? msg : JSON.stringify(msg));
  }
  return res.json() as Promise<PostSummary>;
}

export async function createReply(
  postId: string,
  payload: CreateReplyPayload,
): Promise<PostDetail> {
  const token = await getAccessToken();
  const res = await fetch(
    `${getApiBaseUrl()}/api/community/${postId}/replies`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify(payload),
    },
  );
  if (!res.ok) throw new Error(`createReply failed: ${res.status}`);
  return res.json() as Promise<PostDetail>;
}

export async function toggleUpvote(postId: string): Promise<UpvoteResponse> {
  const token = await getAccessToken();
  const res = await fetch(
    `${getApiBaseUrl()}/api/community/${postId}/upvote`,
    {
      method: "POST",
      headers: { Authorization: `Bearer ${token}` },
    },
  );
  if (!res.ok) throw new Error(`toggleUpvote failed: ${res.status}`);
  return res.json() as Promise<UpvoteResponse>;
}

export async function togglePostDownvote(postId: string): Promise<VoteResponse> {
  const token = await getAccessToken();
  const res = await fetch(
    `${getApiBaseUrl()}/api/community/${postId}/downvote`,
    {
      method: "POST",
      headers: { Authorization: `Bearer ${token}` },
    },
  );
  if (!res.ok) throw new Error(`togglePostDownvote failed: ${res.status}`);
  return res.json() as Promise<VoteResponse>;
}

export async function toggleReplyUpvote(
  postId: string,
  replyId: string,
): Promise<VoteResponse> {
  const token = await getAccessToken();
  const res = await fetch(
    `${getApiBaseUrl()}/api/community/${postId}/replies/${replyId}/upvote`,
    {
      method: "POST",
      headers: { Authorization: `Bearer ${token}` },
    },
  );
  if (!res.ok) throw new Error(`toggleReplyUpvote failed: ${res.status}`);
  return res.json() as Promise<VoteResponse>;
}

export async function toggleReplyDownvote(
  postId: string,
  replyId: string,
): Promise<VoteResponse> {
  const token = await getAccessToken();
  const res = await fetch(
    `${getApiBaseUrl()}/api/community/${postId}/replies/${replyId}/downvote`,
    {
      method: "POST",
      headers: { Authorization: `Bearer ${token}` },
    },
  );
  if (!res.ok) throw new Error(`toggleReplyDownvote failed: ${res.status}`);
  return res.json() as Promise<VoteResponse>;
}

export async function getNotifications(): Promise<NotificationOut[]> {
  const token = await getAccessToken();
  const res = await fetch(`${getApiBaseUrl()}/api/community/notifications`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  if (!res.ok) throw new Error(`getNotifications failed: ${res.status}`);
  return res.json() as Promise<NotificationOut[]>;
}

export async function markNotificationsRead(): Promise<void> {
  const token = await getAccessToken();
  await fetch(`${getApiBaseUrl()}/api/community/notifications/read`, {
    method: "POST",
    headers: { Authorization: `Bearer ${token}` },
  });
}

File: services/api/app/routers/community.py
from typing import Annotated, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status

from app.auth.deps import get_current_user_access
from app.db.client import get_supabase_for_access_token
from app.db.community import (
    create_community_post,
    create_community_reply,
    get_community_post_with_replies,
    get_departments,
    get_notifications,
    get_user_posts,
    list_community_posts,
    mark_notifications_read,
    toggle_post_downvote,
    toggle_reply_downvote,
    toggle_reply_upvote,
    toggle_upvote,
)
from app.models.community import (
    CreatePostRequest,
    CreateReplyRequest,
    NotificationOut,
    PostDetail,
    PostListResponse,
    PostSummary,
    UpvoteResponse,
    VoteResponse,
)

router = APIRouter(prefix="/community", tags=["community"])


@router.get("", response_model=PostListResponse)
def list_posts(
    course_code: Optional[str] = Query(default=None),
    professor_name: Optional[str] = Query(default=None),
    search: Optional[str] = Query(default=None),
    sort_by: str = Query(default="newest"),
    department: Optional[str] = Query(default=None),
    course_number: Optional[str] = Query(default=None),
    page: int = Query(default=1, ge=1),
    auth: tuple[str, str] = Depends(get_current_user_access),
) -> PostListResponse:
    _, access_token = auth
    client = get_supabase_for_access_token(access_token)
    return list_community_posts(
        client,
        course_code=course_code,
        professor_name=professor_name,
        search=search,
        sort_by=sort_by,
        department=department,
        course_number=course_number,
        page=page,
    )


@router.post("", response_model=PostSummary, status_code=status.HTTP_201_CREATED)
def create_post(
    body: CreatePostRequest,
    auth: tuple[str, str] = Depends(get_current_user_access),
) -> PostSummary:
    user_id, access_token = auth
    client = get_supabase_for_access_token(access_token)
    try:
        return create_community_post(
            client,
            user_id,
            title=body.title,
            body=body.body,
            course_code=body.course_code,
            professor_name=body.professor_name,
            is_anonymous=body.is_anonymous,
            general_tags=body.general_tags,
        )
    except Exception as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


# NOTE: /notifications, /departments, and /notifications/read must be declared BEFORE /{post_id}
# so FastAPI does not try to match them as a post_id path param.

@router.get("/departments", response_model=list[str])
def list_departments(
    auth: tuple[str, str] = Depends(get_current_user_access),
) -> list[str]:
    _, access_token = auth
    client = get_supabase_for_access_token(access_token)
    return get_departments(client)


@router.get("/notifications", response_model=list[NotificationOut])
def list_notifications(
    auth: tuple[str, str] = Depends(get_current_user_access),
) -> list[NotificationOut]:
    user_id, access_token = auth
    client = get_supabase_for_access_token(access_token)
    return get_notifications(client, user_id)


@router.post("/notifications/read", status_code=status.HTTP_204_NO_CONTENT)
def read_notifications(
    auth: tuple[str, str] = Depends(get_current_user_access),
) -> None:
    user_id, access_token = auth
    client = get_supabase_for_access_token(access_token)
    mark_notifications_read(client, user_id)


@router.get("/{post_id}", response_model=PostDetail)
def get_post(
    post_id: str,
    auth: tuple[str, str] = Depends(get_current_user_access),
) -> PostDetail:
    _, access_token = auth
    client = get_supabase_for_access_token(access_token)
    post = get_community_post_with_replies(client, post_id)
    if post is None:
        raise HTTPException(status_code=404, detail="Post not found")
    return post


@router.post(
    "/{post_id}/upvote",
    response_model=UpvoteResponse,
    status_code=status.HTTP_200_OK,
)
def upvote_post(
    post_id: str,
    auth: tuple[str, str] = Depends(get_current_user_access),
) -> UpvoteResponse:
    user_id, access_token = auth
    client = get_supabase_for_access_token(access_token)
    try:
        upvoted, count = toggle_upvote(client, post_id, user_id)
        return UpvoteResponse(upvoted=upvoted, upvote_count=count)
    except Exception as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.post(
    "/{post_id}/downvote",
    response_model=VoteResponse,
    status_code=status.HTTP_200_OK,
)
def downvote_post(
    post_id: str,
    auth: tuple[str, str] = Depends(get_current_user_access),
) -> VoteResponse:
    user_id, access_token = auth
    client = get_supabase_for_access_token(access_token)
    try:
        voted, up_count, down_count = toggle_post_downvote(client, post_id, user_id)
        return VoteResponse(voted=voted, upvote_count=up_count, downvote_count=down_count)
    except Exception as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.post(
    "/{post_id}/replies",
    response_model=PostDetail,
    status_code=status.HTTP_201_CREATED,
)
def create_reply(
    post_id: str,
    body: CreateReplyRequest,
    auth: tuple[str, str] = Depends(get_current_user_access),
) -> PostDetail:
    user_id, access_token = auth
    client = get_supabase_for_access_token(access_token)
    try:
        create_community_reply(
            client,
            user_id,
            post_id,
            body=body.body,
            parent_reply_id=body.parent_reply_id,
            is_anonymous=body.is_anonymous,
        )
    except LookupError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except Exception as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

    post = get_community_post_with_replies(client, post_id)
    if post is None:
        raise HTTPException(status_code=404, detail="Post not found")
    return post


@router.post(
    "/{post_id}/replies/{reply_id}/upvote",
    response_model=VoteResponse,
    status_code=status.HTTP_200_OK,
)
def upvote_reply(
    post_id: str,
    reply_id: str,
    auth: tuple[str, str] = Depends(get_current_user_access),
) -> VoteResponse:
    user_id, access_token = auth
    client = get_supabase_for_access_token(access_token)
    try:
        voted, up_count, down_count = toggle_reply_upvote(client, reply_id, user_id)
        return VoteResponse(voted=voted, upvote_count=up_count, downvote_count=down_count)
    except Exception as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.post(
    "/{post_id}/replies/{reply_id}/downvote",
    response_model=VoteResponse,
    status_code=status.HTTP_200_OK,
)
def downvote_reply(
    post_id: str,
    reply_id: str,
    auth: tuple[str, str] = Depends(get_current_user_access),
) -> VoteResponse:
    user_id, access_token = auth
    client = get_supabase_for_access_token(access_token)
    try:
        voted, up_count, down_count = toggle_reply_downvote(client, reply_id, user_id)
        return VoteResponse(voted=voted, upvote_count=up_count, downvote_count=down_count)
    except Exception as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
